### ISSUE TITLE
Postgres Crash when using PLV8

### ISSUE THREAD

**Issue by mchhatani** [plv8/plv8#291]
[postgres_10_4_crash_logs.txt](https://github.com/plv8/plv8/files/2120884/postgres_10_4_crash_logs.txt)
[postgres_10_4_crash_schema.txt](https://github.com/plv8/plv8/files/2120887/postgres_10_4_crash_schema.txt)

I am trying to use the PLV8 via function and while using the function created via PLV8 in one of the create materialized view, postgres crashes, attached is the log file with DEBUG5 turned on.

SQL which is breaking the code and SQL function is attached.

Creating materialized view - mat_view_by_product - is the one which causes the crash.

I have tested the same in below environments

1. Test 1 - Passed
Mac , Postgres - 10.4, PLV8 - 2.1.0

2. Test 2 - Passed
AWS RDS , Postgres - 9.6.6 , PLV8 - 1.5.0

3. Test 3 - Fail, This test was passing earlier on 10.3 but fails on 10.4
AWS EC2 , Postgres - 10.4, PLV8 - 2.3.4

4. Test 4 - Fail
AWS RDS , Postgres - 10.3 , PLV8 - 2.1.0

Please let me know if any more information is required to assist in this problem.

**Comment by JerrySievert** [comment]
can you break it down to a single function that causes the issue?

**Comment by mchhatani** [comment]
Hello Jerry,

Below is the SQL which causes the issue

```
CREATE MATERIALIZED VIEW mat_view_by_product AS
SELECT
    facility_alias_id,
    group_type_id,
    product_id,
    po_date_month,
    histogram_agg(histogram) AS histogram,
    sum(total_spend) AS total_spend,
    min(min_price) AS min_price,
    max(max_price) AS max_price,
    min(min_po_date) AS min_po_date,
    max(max_po_date) AS max_po_date,
    (array_agg(most_recent_price ORDER BY max_po_date))[1] AS most_recent_price,
    sum(total_eaches) AS total_eaches
FROM
    mat_view_by_catalog
GROUP BY
    facility_alias_id,
    group_type_id,
    product_id,
    po_date_month;
```

**Comment by JerrySievert** [comment]
unfortunately, that doesn't help very much.

the logs you provided only show that a crash occurred during a function that uses plv8, not that plv8 caused the crash, nor where in the plv8 (or v8) code that the crash occurred, and you don't have data that would also cause this.

if you can create a test case that includes a simple function that causes the crash, along with some data that causes it, I can help track down the issue.

**Comment by mchhatani** [comment]
I am creating a dataset and it will be around 500MB to reproduce the issue because without the data issue is not reproducible, how can I share the data.

**Comment by mchhatani** [comment]
Thanks Jerry for the feedback, I was actually able to just create a dump of 12MB data, so you just load the data and then execute below SQL to reproduce the issue

```
SELECT
    facility_alias_id,
    group_type_id,
    product_id,
    po_date_month,histogram_agg(histogram) AS histogram,
    sum(total_spend) AS total_spend,
    min(min_price) AS min_price,
    max(max_price) AS max_price,
    min(min_po_date) AS min_po_date,
    max(max_po_date) AS max_po_date,
    (array_agg(most_recent_price ORDER BY max_po_date))[1] AS most_recent_price,
    sum(total_eaches) AS total_eaches
FROM
    test_data
GROUP BY
    facility_alias_id,
    group_type_id,
    product_id,
    po_date_month;
```
[postgres_10_4_plv8_crash.txt](https://github.com/plv8/plv8/files/2121504/postgres_10_4_plv8_crash.txt)

**Comment by JerrySievert** [comment]
I'm able to get an occasional crash in `palloc()` (Postgres's memory allocation function) under MacOS, but not ubuntu.  still trying to track down what in `palloc` could be crashing.

**Comment by mchhatani** [comment]
I did below test

Scenario 1

1. Mac OS - There was no issues
```
procured=# select version();
                                                    version                                                    
------------------------------------------------------------------------------------------------------
 PostgreSQL 10.4 on x86_64-apple-darwin17.4.0, compiled by Apple LLVM version 9.1.0 (clang-902.0.39.1), 64-bit
test_crash_plv8=# select * from pg_extension ;
 extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition 
---------+----------+--------------+----------------+------------+-----------+--------------
 plpgsql |       10 |           11 | f              | 1.0        |           | 
 plv8    |       10 |           11 | f              | 2.1.0      |           | 
```

2. AWS EC2 (AWS AMI /AWS Linux) - Issue happened multiple times
-bash-4.2$ uname -a
Linux ip-10-2-4-41 4.14.33-51.37.amzn1.x86_64 #1 SMP Thu May 3 20:07:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
test_plv8_crash=# select version();
                                                 version                                                 
------------------------------------------------------------------------------------------------------
 PostgreSQL 10.4 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-18), 64-bit
 extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition 
---------+----------+--------------+----------------+------------+-----------+--------------
 plpgsql |       10 |           11 | f              | 1.0        |           | 
 plv8    |       10 |           11 | f              | 2.3.4      |           | 
```

3. AWS RDS - 9.6.6 - No issues
```
test_crash_plv8=> select version();
                                                 version                                                 
---------------------------------------------------------------------------------------------------------
 PostgreSQL 9.6.6 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.3 20140911 (Red Hat 4.8.3-9), 64-bit
(1 row)

test_crash_plv8=> select * from pg_extension ;
 extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition 
---------+----------+--------------+----------------+------------+-----------+--------------
 plpgsql |       10 |           11 | f              | 1.0        |           | 
 plv8    |       10 |           11 | f              | 1.5.3      |           | 
```

I am still working to test this on below machine configuration
```
postgres=> select version();
                                                                  version                                                                   
--------------------------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 10.4 (Ubuntu 10.4-2.pgdg16.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609, 64-bit
(1 row)

postgres=> \q
mchhatani@devops:~/plv8-2.3.4$ uname -a
Linux ip-10-10-2-51 4.4.0-1055-aws #64-Ubuntu SMP Thu Apr 5 17:06:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by JerrySievert** [comment]
I have it traced to a bad allocation in postgres (pg 10.3), I'm just trying to figure out *why* it's a bad allocation in postgres - it appears that the `context` is invalid, but it shouldn't be given that it's the "top" context.

**Comment by mchhatani** [comment]
Thanks Jerry for working on this, I already reported this issue to Postgres Mailing List - here is the description , if you want to add more comments there ->  Postgres 10.4 crashing when using PLV8

I was told that to report it here and than you can report it to them if this looks like core postgres issue.

**Comment by JerrySievert** [comment]
I'm on the postgresql irc channel now - since the `TopMemoryContext` is corrupted somehow:

```
(lldb) p *TopMemoryContext->methods
(MemoryContextMethods) $4 = {
  alloc = 0x00000000000000d1
  free_p = 0x00007f9b12402470
  realloc = 0x0000000000000001
  init = 0x00007f9b1284a358
  reset = 0x0000000000000000
  delete_context = 0x0000000000000000
  get_chunk_space = 0x00007f9b12402538
  is_empty = 0x0000000000000000
  stats = 0x00007f9b12848600
}
```

**Comment by mchhatani** [comment]
Thanks Jerry for the update, I will continue with my testing on couple of other platforms and update you.

**Comment by JerrySievert** [comment]
I'm going to have v8 use its own default array buffer allocation manager, instead of having it try to use pg's.

this appears to fix the problem, by avoiding it completely.  I'm waiting for ci to pass, and will put out a new release.

**Comment by mchhatani** [comment]
Thanks Jerry for the update, let me know once new release is out I will be ready to test it.

**Comment by JerrySievert** [comment]
it's merged into `master` if you want to do a `git clone` and try that.

**Comment by mchhatani** [comment]
Build in progress, will update you once I can test it.

**Comment by mchhatani** [comment]
Jerry, it looks like issue does not happen anymore atleast with the 12 MB data I have provider and the version is 2.3.6, I will do some more extensive testing and let you know if I find any more issues.


I really appreciate your help on this issue.

**Comment by mchhatani** [comment]
Jerry, 

Just wanted to check why this issue is not happening in 1.5.3 and 2.1.0 with 9.6.6 , since that code does not have the fix, so could this be because of something in postgres10.

**Comment by mchhatani** [comment]
FYI... This issue happens on 9.6.8 also.

**Comment by JerrySievert** [comment]
yes, I'm aware. as I said, I was able to ascertain the cause of the problem and have released an update to `2.3.6`, and back ported the fix to `2.1.2`.
